### DISCUSSION TITLE
Displaying transaction calldata

### DISCUSSION THREAD

**Discussion by ditto-eth** [paradigmxyz/rivet#26] [Ideas]
Wanted to start a discussion on displaying the tx calldata (since I was working on #15)! This would be part of "View transaction details (including decoded calldata*, logs*, state*, and tracing*)" in the readme

Is there already an easy way to get this? I assume it's normally by verifying the contract, but if this is all your own code, you should be able to read from artifacts/generated files? I did a simple test myself that just copies the selectors and abi from artifacts/wagmi/cli just to debug and use on my own (maybe there's some viem/wagmi utilities already for this though) - it just extracts the 4byte selector, the function interface, and each parameter name/value.

![Screenshot 2023-08-29 at 7 16 57 PM](https://github.com/paradigmxyz/rivet/assets/136378658/49de6e22-f110-4625-92c9-39eeb8165bf0)

**Comment by jxom** [comment]
Looks pretty good! Happy for you to work on this.

Until we have the functionality to extract ABIs from forge artifacts, we can probably just guess the ABI with https://github.com/shazow/whatsabi for now. Let me know if you need any assistance.

**Reply by ditto-eth** [reply]
> Until we have the functionality to extract ABIs from forge artifacts

Awesome, yea what do you think that would look like? ya I wrote a simple util to do translate the 4bytes (basic js to read `methodIdentifiers` in `artifacts/Vm.sol/Vm.json`) by looping through everything in artifacts and creating a json of all the sigs, then generate a ts file and just read it. What would it look like to be able to extract the ABIs otherwise? would it be a viem util? pass a file path? the abis are just at `"abi"` but like ^ I just hacked together using the wagmi cli output

```json
{
  "methodIdentifiers": {
    "activeFork()": "2f103f22",
    "addr(uint256)": "ffa18649",
    ..
  }
}
```

For [whatsabi](https://github.com/shazow/whatsabi), i can figure out how this project does fetch calls I suppose!

**Reply by shazow** [reply]
Ping me if I can help with anything on the whatsabi end of things. :)

The fully offline parameter guessing is still unreliable (going to do a big revamp on this next month possibly), but the selector extraction is quite good and there's various helpers for matching them with 4byte databases.
